File: _work/backend-watchdog-full-repair-02/src/iii-sandbox/packages/sdk/src/sandbox.ts
import type { HttpClient } from "./client.js";
import type {
  SandboxInfo,
  ExecResult,
  ExecStreamChunk,
  SandboxMetrics,
  SnapshotInfo,
} from "./types.js";
import { EnvManager } from "./env.js";
import { FileSystem } from "./filesystem.js";
import { GitManager } from "./git.js";
import { CodeInterpreter } from "./interpreter.js";
import { ProcessManager } from "./process.js";
import { PortManager } from "./port.js";
import { QueueManager } from "./queue.js";
import { StreamManager } from "./stream-manager.js";
import { MonitorManager } from "./monitor.js";
import { TerminalManager } from "./terminal.js";
import { parseExecStream } from "./stream.js";

export class Sandbox {
  readonly env: EnvManager;
  readonly filesystem: FileSystem;
  readonly git: GitManager;
  readonly interpreter: CodeInterpreter;
  readonly processes: ProcessManager;
  readonly ports: PortManager;
  readonly queue: QueueManager;
  readonly streams: StreamManager;
  readonly monitor: MonitorManager;
  readonly terminal: TerminalManager;

  constructor(
    private client: HttpClient,
    public info: SandboxInfo,
  ) {
    this.env = new EnvManager(client, info.id);
    this.filesystem = new FileSystem(client, info.id);
    this.git = new GitManager(client, info.id);
    this.interpreter = new CodeInterpreter(client, info.id);
    this.processes = new ProcessManager(client, info.id);
    this.ports = new PortManager(client, info.id);
    this.queue = new QueueManager(client, info.id);
    this.streams = new StreamManager(client, info.id);
    this.monitor = new MonitorManager(client, info.id);
    this.terminal = new TerminalManager(client, info.id);
  }

  get id(): string {
    return this.info.id;
  }

  get status(): string {
    return this.info.status;
  }

  async exec(command: string, timeout?: number): Promise<ExecResult> {
    return this.client.post<ExecResult>(
      `/sandbox/sandboxes/${this.info.id}/exec`,
      { command, timeout },
    );
  }

  async *execStream(command: string): AsyncGenerator<ExecStreamChunk> {
    const lines = this.client.stream(
      `/sandbox/sandboxes/${this.info.id}/exec/stream`,
      { command },
    );
    yield* parseExecStream(lines);
  }

  async clone(name?: string): Promise<SandboxInfo> {
    return this.client.post<SandboxInfo>(
      `/sandbox/sandboxes/${this.info.id}/clone`,
      { name },
    );
  }

  async pause(): Promise<void> {
    await this.client.post(`/sandbox/sandboxes/${this.info.id}/pause`);
  }

  async resume(): Promise<void> {
    await this.client.post(`/sandbox/sandboxes/${this.info.id}/resume`);
  }

  async kill(): Promise<void> {
    await this.client.del(`/sandbox/sandboxes/${this.info.id}`);
  }

  async metrics(): Promise<SandboxMetrics> {
    return this.client.get<SandboxMetrics>(
      `/sandbox/sandboxes/${this.info.id}/metrics`,
    );
  }

  async snapshot(name?: string): Promise<SnapshotInfo> {
    return this.client.post<SnapshotInfo>(
      `/sandbox/sandboxes/${this.info.id}/snapshots`,
      { name },
    );
  }

  async restore(snapshotId: string): Promise<SandboxInfo> {
    return this.client.post<SandboxInfo>(
      `/sandbox/sandboxes/${this.info.id}/snapshots/restore`,
      { snapshotId },
    );
  }

  async listSnapshots(): Promise<{ snapshots: SnapshotInfo[] }> {
    return this.client.get<{ snapshots: SnapshotInfo[] }>(
      `/sandbox/sandboxes/${this.info.id}/snapshots`,
    );
  }

  async refresh(): Promise<SandboxInfo> {
    const updated = await this.client.get<SandboxInfo>(
      `/sandbox/sandboxes/${this.info.id}`,
    );
    Object.assign(this.info, updated);
    return updated;
  }
}

File: Core kernel Yaiwes/Backend watchdog workflow adaptativo/sandbox-runtime/iii-sandbox/packages/sdk-python/src/iii_sandbox/sandbox.py
from __future__ import annotations

import json
from typing import AsyncGenerator

from .client import HttpClient
from .env import EnvManager
from .filesystem import FileSystem
from .git import GitManager
from .interpreter import CodeInterpreter
from .monitor import MonitorManager
from .port import PortManager
from .process import ProcessManager
from .queue import QueueManager
from .stream import StreamManager
from .types import (
    ExecResult,
    ExecStreamChunk,
    SandboxInfo,
    SandboxMetrics,
    SnapshotInfo,
)


class Sandbox:
    def __init__(self, client: HttpClient, info: SandboxInfo):
        self._client = client
        self.info = info
        self.env = EnvManager(client, info.id)
        self.filesystem = FileSystem(client, info.id)
        self.git = GitManager(client, info.id)
        self.interpreter = CodeInterpreter(client, info.id)
        self.processes = ProcessManager(client, info.id)
        self.ports = PortManager(client, info.id)
        self.queue = QueueManager(client, info.id)
        self.streams = StreamManager(client, info.id)
        self.monitor = MonitorManager(client, info.id)

    @property
    def id(self) -> str:
        return self.info.id

    @property
    def status(self) -> str:
        return self.info.status

    async def exec(
        self, command: str, timeout: int | None = None
    ) -> ExecResult:
        body: dict = {"command": command}
        if timeout is not None:
            body["timeout"] = timeout
        data = await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/exec", body
        )
        return ExecResult(**data)

    async def exec_stream(
        self, command: str
    ) -> AsyncGenerator[ExecStreamChunk, None]:
        lines = self._client.stream(
            f"/sandbox/sandboxes/{self.info.id}/exec/stream",
            {"command": command},
        )
        async for line in lines:
            try:
                chunk = ExecStreamChunk(**json.loads(line))
                yield chunk
                if chunk.type == "exit":
                    return
            except (json.JSONDecodeError, ValueError, TypeError, KeyError):
                continue

    async def clone(self, name: str | None = None) -> SandboxInfo:
        body: dict = {}
        if name is not None:
            body["name"] = name
        data = await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/clone", body
        )
        return SandboxInfo(**data)

    async def pause(self) -> None:
        await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/pause"
        )

    async def resume(self) -> None:
        await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/resume"
        )

    async def kill(self) -> None:
        await self._client.delete(
            f"/sandbox/sandboxes/{self.info.id}"
        )

    async def metrics(self) -> SandboxMetrics:
        data = await self._client.get(
            f"/sandbox/sandboxes/{self.info.id}/metrics"
        )
        return SandboxMetrics(**data)

    async def snapshot(self, name: str | None = None) -> SnapshotInfo:
        body: dict = {}
        if name is not None:
            body["name"] = name
        data = await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/snapshots", body
        )
        return SnapshotInfo(**data)

    async def restore(self, snapshot_id: str) -> SandboxInfo:
        data = await self._client.post(
            f"/sandbox/sandboxes/{self.info.id}/snapshots/restore",
            {"snapshotId": snapshot_id},
        )
        return SandboxInfo(**data)

    async def list_snapshots(self) -> dict:
        return await self._client.get(
            f"/sandbox/sandboxes/{self.info.id}/snapshots"
        )

    async def refresh(self) -> SandboxInfo:
        data = await self._client.get(
            f"/sandbox/sandboxes/{self.info.id}"
        )
        updated = SandboxInfo(**data)
        self.info = updated
        return updated
